src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -612,11 +612,24 @@ void AtlasEngine::_resolveFontMetrics(const wchar_t* requestedFaceName, const Fo
     const auto descentInPx = static_cast<double>(metrics.descent) * designUnitsPerPx;
     const auto lineGapInPx = static_cast<double>(metrics.lineGap) * designUnitsPerPx;
     const auto advanceWidthInPx = static_cast<double>(glyphMetrics.advanceWidth) * designUnitsPerPx;
+    const auto underlineOffsetInPx = static_cast<double>(-metrics.underlinePosition) * designUnitsPerPx;
+    const auto underlineThicknessInPx = static_cast<double>(metrics.underlineThickness) * designUnitsPerPx;
+    const auto strikethroughOffsetInPx = static_cast<double>(-metrics.strikethroughPosition) * designUnitsPerPx;
+    const auto strikethroughThicknessInPx = static_cast<double>(metrics.strikethroughThickness) * designUnitsPerPx;
+    const auto lineThicknessInPx = std::round(std::max(1.0, std::min(underlineThicknessInPx, strikethroughThicknessInPx)));
 
     const auto halfGapInPx = lineGapInPx / 2.0;
     const auto baseline = std::ceil(ascentInPx + halfGapInPx);
+    const auto underlinePosInPx = std::ceil(baseline + underlineOffsetInPx - lineThicknessInPx / 2.0);
+    const auto strikethroughPosInPx = std::round(baseline + strikethroughOffsetInPx - lineThicknessInPx / 2.0);
+    const auto cellHeightViaDescent = std::ceil(baseline + descentInPx + halfGapInPx);
+    // A double-underline is 3 lines tall due to a 1 line gap between the two lines.
+    // This logic should be kept in sync with AtlasEngine::_updateConstantBuffer,
+    // which calculates the offset/position of the lower line.
+    const auto cellHeightViaUnderlines = underlinePosInPx + 3 * lineThicknessInPx;

```

Comment:
With Consolas for instance the `cellHeightViaUnderlines` is usally larger than the `cellHeightViaDescent`. If we didn't make this larger change here, the double-underline would appear clipped off.

Reply:
Have you considered just using the same algorithm that we have in the other renderers? Because people can be a picky about line spacing, and I suspect some users will not appreciate having their line height stretched to accommodate a double underline that they may never even use.

And eventually I'm assuming we'll have options for adjust line spacing (e.g. #3498), so we really need to be thinking in terms of fitting the content within the height that the user as requested (at least as best we can) rather than adjusting the height to fit the content.

Reply:
Yeah I have. I choose to not use that, because I noticed that for most common fonts the descender height is approximately the same as the underline offset + 3 line widths down to just +/- 1px for fonts I have installed.

Since the difference was so small I built the PR this way. It allows me align double underlines with regular underlines in a consistent manner:

Monofur 24pt 150%
![image](https://user-images.githubusercontent.com/2256941/180894962-142f26f4-0399-4d70-960d-0a9a041bd25f.png)

Consolas 24pt 100%
![image](https://user-images.githubusercontent.com/2256941/180895079-48feef40-a13f-43ad-9e73-0c917dfdc32a.png)

MS Gothic 10pt 100%
![image](https://user-images.githubusercontent.com/2256941/180895162-16ad9b8e-7537-4619-8755-a8cecef0ba8c.png)

I liked that result a lot! And it only costs us +1px line height at most (for fonts I tested). Of course it's possible that it can increase the line height in weird ways for certain fonts, but so far I'm feeling somewhat confident that we should try this first.
Do you think that's acceptable? I'm open to adopting the algorithm you used in GdiEngine, but I feel like this approach has its advantages.

#3498 is a difficult one IMO. If we were to treat it like CSS `line-height`, then glyphs would simply overlap each other and not shrink the glyphs themselves or their underlines. I'd strongly prefer such a behavior if anything over fitting text contents into the given height. (I know how to do that, but it's "a bit" involved.)
A simpler solution is of course to simply clip the cell size to the given size and downscale glyphs as needed and is what I intend as a stop-gap solution as well. However I'm not really concerned about underlines being clipped away - if the user specifies a line height that's too small to support descenders (= underline offset + 3 lines widths +/- 1px), I'd simply clip the underline off. In practice that probably even works for most users. 🤔

Reply:
I know when I was testing I definitely had fonts where the underline position was right up against the bottom of the cell, which is why I had the fallback option of rendering the second line above the first. So even with a single pixel line width you'd need to stretch the height by at least two pixels on those cases, and wider line widths would assumedly be even more.

That said, I don't feel strongly about this either way. I only brought it up because I know some people are quite fanatical about line heights. Also, if we're hoping to replace the GDI engine in conhost with Atlas one day, I suspect we might need to retain the current cell height calculations for backwards compatibility reasons. I may be wrong about that though.

As for #3498, I could understand having the glyphs overlapping each other if the line height was too small for them to fit, but I'd definitely expect an overline to be clamped to the top of the cell, and the underline to be clamped to the bottom. To me it makes more sense to have an underline overlapping its own text rather than the text of the row below it.

Reply:
Actually, checking Word for reference, it seems that putting the second line above the underline might be generally preferable (Consolas 24pt):
![image](https://user-images.githubusercontent.com/2256941/180903923-16ebb4b0-c166-40b3-93b4-1cff3dfac682.png)

It's halfway between baseline and underline. Interesting! But I'm not sure if I'm ready for such a more complex approach... I'd rather adopt the Gdi/DxEngine code and iterate on this later on.

Reply:
That's interesting. It also looks to me like they've taken the underline and sort of split it in half to produce the double underline, which makes it much more compact. But that only really works when the base underline width is more than a pixel to begin with.

Reply:
So I did a little something... 😎 It's just 6 lines of code, but the result is surprisingly close to what Word does. It also adapts to small descender heights if needed.

Consolas 24pt 150%
![image](https://user-images.githubusercontent.com/2256941/181041449-7712c813-80b1-434e-9b62-5d506cdba529.png)

Courier New 24pt
![image](https://user-images.githubusercontent.com/2256941/181040651-ee4a76aa-45c8-4ee4-acbf-aa48bd1c97bf.png)

Cascadia Code 24pt 150%
![image](https://user-images.githubusercontent.com/2256941/181045138-920677e6-bc92-47fd-a99c-f7ab83dc78fe.png)

All images are at 150% display scale.